aperture.go

```diff
@@ -231,6 +186,70 @@ func setupLogging(cfg *config) error {
 	return build.ParseAndSetDebugLevels(cfg.DebugLevel, logWriter)
 }
 
+// getTLSConfig returns a TLS configuration for either a self-signed certificate
+// or one obtained through Let's Encrypt.
+func getTLSConfig(serverName string, autoCert bool) (*tls.Config, error) {
+	var tlsKeyFile, tlsCertFile string
+	switch {
+	// When using autocert, we set a TLSConfig on the server so the key and
+	// cert file we pass in are ignored and don't need to exist.
+	case autoCert:
+		serverName := serverName
+		if serverName == "" {
+			return nil, fmt.Errorf("servername option is " +
+				"required for secure operation")
+		}
+
+		certDir := filepath.Join(apertureDataDir, "autocert")
+		log.Infof("Configuring autocert for server %v with cache dir "+
+			"%v", serverName, certDir)
+
+		manager := autocert.Manager{
+			Cache:      autocert.DirCache(certDir),
+			Prompt:     autocert.AcceptTOS,
+			HostPolicy: autocert.HostWhitelist(serverName),
+		}
+
+		go func() {
+			err := http.ListenAndServe(
+				":http", manager.HTTPHandler(nil),
+			)
+			if err != nil {
+				log.Errorf("autocert http: %v", err)
+			}
+		}()
+		return &tls.Config{
+			GetCertificate: manager.GetCertificate,
+		}, nil
+
+	// If we're not using autocert, we want to create self-signed TLS certs
+	// and save them at the specified location (if they don't already
+	// exist).
+	default:
+		tlsKeyFile = filepath.Join(apertureDataDir, defaultTLSKeyFilename)
+		tlsCertFile = filepath.Join(apertureDataDir, defaultTLSCertFilename)
+		if !fileExists(tlsCertFile) && !fileExists(tlsKeyFile) {
+			log.Infof("Generating TLS certificates...")
+			err := cert.GenCertPair(
+				"aperture autogenerated cert", tlsCertFile,
+				tlsKeyFile, nil, nil,
+				cert.DefaultAutogenValidity,
+			)
+			if err != nil {
+				return nil, err
+			}
+			log.Infof("Done generating TLS certificates")
+		}
+
+		certData, _, err := cert.LoadCert(tlsCertFile, tlsKeyFile)
+		if err != nil {
+			return nil, err
+		}
+
+		return cert.TLSConfFromCert(certData), nil
```

Comment:
This change is more than only a function extraction, probably deserves a mention in the commit message. Or a separate commit doing this first and the next commit extracting the function.